s3transfer/upload.py

```diff
@@ -286,6 +286,157 @@ def _get_put_object_fileobj(self, fileobj):
         return fileobj
 
 
+class UploadNonSeekableInputManager(UploadInputManager):
+    """Upload utility for a file-like object that cannot seek."""
+    def __init__(self, osutil, transfer_coordinator):
+        super(UploadNonSeekableInputManager, self).__init__(
+            osutil, transfer_coordinator)
+        self._initial_data = None
+
+    @classmethod
+    def is_compatible(cls, upload_source):
+        return not seekable(upload_source)
+
+    def stores_body_in_memory(self, operation_name):
+        return True
+
+    def provide_transfer_size(self, transfer_future):
+        # No-op because there is no way to do this short of reading the entire
+        # body into memory.
+        return
+
+    def requires_multipart_upload(self, transfer_future, config):
+        # This is tricky to determine in this case because we can't know how
+        # large the input is. So to figure it out, we read data into memory
+        # up until the threshold and compare how much data was actually read
+        # against the threshold.
+        fileobj = transfer_future.meta.call_args.fileobj
+        threshold = config.multipart_threshold
+
+        # Check to see if any data has been read in already.
+        if self._initial_data is not None:
+            current_data_length = len(self._initial_data)
+        else:
+            self._initial_data = b''
+            current_data_length = 0
+
+        # If the amount of data already read is less than the current
+        # threshold, then we attempt to read in the difference.
+        data_to_read = threshold - current_data_length
+        if data_to_read > 0:
+            data = fileobj.read(data_to_read)
+            self._initial_data += data
+
+        if len(self._initial_data) < threshold:
+            return False
+        else:
+            return True
+
+    def get_put_object_body(self, transfer_future):
+        callbacks = get_callbacks(transfer_future, 'progress')
+        fileobj = transfer_future.meta.call_args.fileobj
+
+        if self._initial_data is None:
+            data = b''
+        else:
+            data = self._initial_data
+        body = self._wrap_part_data(data + fileobj.read(), callbacks)
+
+        # Zero out the stored data so we don't have additional copies
+        # hanging around in memory.
+        self._initial_data = None
+        return body
+
+    def yield_upload_part_bodies(self, transfer_future, config):
+        part_size = config.multipart_chunksize
+        file_object = transfer_future.meta.call_args.fileobj
+        callbacks = get_callbacks(transfer_future, 'progress')
+        part_number = 0
+
+        # First go over the parts generated by calling
+        # `requires_multipart_upload` if any.
+        if self._initial_data is not None:
+            parts = self._yield_first_parts(transfer_future, config, callbacks)
+            for part in parts:
+                part_number += 1
+                yield part_number, part
```

Comment:
We talked about this but it would be great if we had all of this logic under one loop and one yield from. The two seperate loops and the three places to yield from is a bit complicated.